pkg/server/plugin/upstreamauthority/gcpcas/gcpcas.go

```diff
@@ -0,0 +1,393 @@
+	createRequest := privatecapb.CreateCertificateRequest{
+		Parent: parentPath,
+		// https://pkg.go.dev/google.golang.org/genproto/googleapis/cloud/security/privateca/v1beta1#Certificate
+		Certificate: &privatecapb.Certificate{
+			Lifetime: durationpb.New(validity),
+			// https://pkg.go.dev/google.golang.org/genproto/googleapis/cloud/security/privateca/v1beta1#CertificateConfig
+			CertificateConfig: &privatecapb.Certificate_Config{
+				Config: &privatecapb.CertificateConfig{
+					PublicKey: &privatecapb.PublicKey{
+						Type: privatecapb.PublicKey_PEM_EC_KEY,
+						Key: pem.EncodeToMemory(
+							&pem.Block{
+								Type:  publicKeyType,
+								Bytes: csrParsed.RawSubjectPublicKeyInfo,
+							},
+						),
+					},
+					SubjectConfig: &privatecapb.CertificateConfig_SubjectConfig{
+						Subject:        &subject,
+						CommonName:     csrParsed.Subject.CommonName,
+						SubjectAltName: &san,
+					},
+					// https://cloud.google.com/certificate-authority-service/docs/reusable-configs#mutual_tls_w_path_length_0
+					// https://cloud.google.com/sdk/gcloud/reference/beta/privateca/roots/create
+					// https://pkg.go.dev/google.golang.org/genproto/googleapis/cloud/security/privateca/v1beta1#ReusableConfigWrapper
+					// https://pkg.go.dev/google.golang.org/genproto/googleapis/cloud/security/privateca/v1beta1#ReusableConfigWrapper_ReusableConfig
+					ReusableConfig: &privatecapb.ReusableConfigWrapper{
+						ConfigValues: &privatecapb.ReusableConfigWrapper_ReusableConfig{
+							ReusableConfig: fmt.Sprintf("projects/privateca-data/locations/%s/reusableConfigs/subordinate-mtls-pathlen-0", config.RootSpec.Location),
+						},
+					},
+				},
+			},
+		},
+	}
+
+	cresp, err := pcaClient.CreateCertificate(ctx, &createRequest)
+	if err != nil {
+		return nil, err
+	}
+	if len(cresp.PemCertificateChain) == 0 {
+		return nil, makeError(codes.Internal, "got no certificates in the chain")
+	}
+
+	cert, err := pemutil.ParseCertificate([]byte(cresp.GetPemCertificate()))
+	if err != nil {
+		return nil, err
+	}
+
+	certChain := make([]*x509.Certificate, len(cresp.PemCertificateChain))
+	for i, c := range cresp.PemCertificateChain {
+		certChain[i], err = pemutil.ParseCertificate([]byte(c))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// All else comprises the chain (including the issued certificate)
+	// We don't include the root, since we pack that into the trust bundle.
+	fullChain := []*x509.Certificate{cert}
+	fullChain = append(fullChain, certChain[:len(certChain)-1]...)
+
+	derChain := x509util.RawCertsFromCertificates(fullChain)
+
+	// Then we append all the extra cert roots we loaded, plus anything we loaded from the
+	// trust bundle spec
+	rootBundle := make([]*x509.Certificate, 0, len(trustBundle))
+	for _, c := range trustBundle {
+		// Note. We are intentionally retrieving the immediate CAs from the GCP CAS that
```

Comment:
Does this mean that we can end up with intermediates in the trust bundle that don't chain up to a root also in the trust bundle? This can be very problematic for downstream software, which does not always verify such chains successfully by default (e.g. OpenSSL).

Reply:
The scenario that I have in mind is:

Lets say that there are three departments of the same organization in google cloud project. The three departments have their own CA - say caX, caY and caZ respectively. And they might all have the same root CA - caM

If each department is their own trust domain ( say domain X, Y and Z ) and trust domain X and Y needs to trust each other but not domain Z

If we have caM in the trust bundle, then wouldn't it allow any department to accept any SVID generated from any department?

I was hoping that if we can tag with appropriate labels, then the trust bundle will only contain caX and caY rather than caM

Or is it a precondition that each trust domain should have an exclusive root CA? That would also eliminate the above problem

Reply:
Outside of verification woes by stacks that use OpenSSL (and probably others), I think including intermediates in the trust bundle is dangerous. It makes it much easier to accidentally open up trust in unintended ways. In this case, what you REALLY trust is the root. Without some way of constraining that intermediate, its hard to anchor trust in it. What you don't want is workloads in two trust domains with intermediates in the trust bundle to accidentally trust each other because they got their hands on the root somewhere (say from the system pool) and included it in their verification chain. That opens up doors for one trust domain to impersonate workloads in the other trust domain.

Reply:
> Or is it a precondition that each trust domain should have an exclusive root CA? That would also eliminate the above problem

It's not strictly required, but is strongly encouraged and generally assumed

I agree with everything @azdagron said. SPIRE used to include intermediates in the bundle (and exclude the real roots) to address the use case you're describing, but we found that in practice it didn't work well, wasn't very safe, and led to a lot of confusion. Removing certs from the bundle is hard from a compat standpoint - adding them is of course easy :)

I think people that find themselves in the circumstance you describe... which I might summarize as "We all have to have intermediates that share a root due to corporate policy, but we don't _really_ trust each other" ... should lean on name constraints to solve their problems.

Reply:
Makes sense. Thank you. Switched to root rather than intermediate